nethttp/server.go

```diff
@@ -129,13 +129,15 @@ func MiddlewareFunc(tr opentracing.Tracer, h http.HandlerFunc, options ...MWOpti
 		sct := &statusCodeTracker{w, 200}
 		r = r.WithContext(opentracing.ContextWithSpan(r.Context(), sp))
 
-		h(sct.wrappedResponseWriter(), r)
+		defer func() {
+			ext.HTTPStatusCode.Set(sp, uint16(sct.status))
+			if sct.status >= http.StatusInternalServerError {
```

Comment:
We will (I think) return 200 with no error in the case where the code has panicked. This could potentially be a little misleading.

We could consider using

 `if recover() != nil`

to detect when we have panicked. We could then set Error=>true and potentially also update the response code. This would mean we distinguish between the panic and non-panic case.

Reply:
Yeah! I've forgotten the problem of status code 200 after panicked. After thinking of this, I found it's difficult to find a best solution, any of proposals below has some drawback.

**Proposal 1**
As your suggestion like below:
```go
       defer func() {                                                          
            if recover() != nil {                                                                            
                ext.HTTPStatusCode.Set(                                         
                    sp, uint16(http.StatusInternalServerError))                 
            } else {                                                                                            
                ext.HTTPStatusCode.Set(sp, uint16(sct.status))           
            }                                                                                                       
            sp.Finish()                                                         
        }()  
```
But if we use `recover()`, we would change the behavior here, that means no panic would be thrown out.

**Proposal 2**
```go
       defer func() {
            var err error                                                          
            if err = recover(); err != nil {                                                                            
                ext.HTTPStatusCode.Set(                                         
                    sp, uint16(http.StatusInternalServerError))
                ext.Error.Set(sp, true)                 
            } else {                                                                                            
                ext.HTTPStatusCode.Set(sp, uint16(sct.status))           
            }                                                                                                       
            sp.Finish()
            if err != nil {
                panic(err)
            }                                                      
        }()  
```
If we `panic` again after `recover`, the inner panic msg would be thrown out, but the stacktrace of panic would be changed... 

**Proposal 3**
```
		sct := &statusCodeTracker{w, 500}
		r = r.WithContext(opentracing.ContextWithSpan(r.Context(), sp))

		defer func() {
			ext.HTTPStatusCode.Set(sp, uint16(sct.status))
			if sct.status >= http.StatusInternalServerError {
				ext.Error.Set(sp, true)
			}
			sp.Finish()
		}()
```
Just change the default status code to 500 of statusCodeTracker. 
* If inner http handler panicked before `WriteHeader` been called, the default status code 500 would be used. That is what we want.
* If panic happens after `WriteHeader` been called, `ext.HTTPStatusCode.Set(sp, uint16(sct.status))` still be correct, but `ext.Error.Set(sp, true)` should be executed when ever `if sct.status >= http.StatusInternalServerError` is true or false. The proposal is not met our expectation.

It's difficult... any ideas?